dask/order.py

```diff
@@ -202,125 +193,212 @@ def dependencies_key(x):
-
-                    result[item] = i
-                    i += 1
-                    deps = dependents[item]
-
-                    if metrics[item][3] == 2:  # min_height
-                        # Don't leave any dangling single nodes!  Finish all dependents that are
-                        # ready and are also root nodes.  Doing this here also lets us continue
-                        # down to a different root node.
-                        finish_now = {
-                            dep
-                            for dep in deps
-                            if not dependents[dep] and num_needed[dep] == 1
-                        }
-                        if finish_now:
-                            deps -= finish_now
-                            if len(finish_now) > 1:
-                                finish_now = sorted(finish_now, key=dependents_key)
-                            for dep in finish_now:
-                                result[dep] = i
-                                i += 1
-
-                    if deps:
+                            later_nodes[key] = [deps]
+                    else:
+                        dep_pools = {}
                         for dep in deps:
-                            num_needed[dep] -= 1
-                        if not inner_stack:
-                            # Continue towards our tactical goal (an easy-to-compute root)
-                            if len(deps) == 1:
-                                inner_stack_extend(deps)
+                            key = partition_keys[dep]
+                            if key in dep_pools:
+                                dep_pools[key].append(dep)
                             else:
-                                # If there are many, many dependents, then calculating
-                                # `dependents_key` here could be relatively expensive.
-                                # However, this is still probably worth it tactically.
-                                inner_stack_append(min(deps, key=dependents_key))
-                                outer_stack_seeds_append(item)
+                                dep_pools[key] = [dep]
+                        item_key = partition_keys[item]
+                        if inner_stack:
+                            prev_key = partition_keys[inner_stack[0]]
+                            now_keys = []  # < inner_stack[0]
+                            for key, vals in dep_pools.items():
+                                if key < prev_key:
+                                    now_keys.append(key)
+                                elif key < item_key:
+                                    if key in next_nodes:
+                                        next_nodes[key].append(vals)
+                                    else:
+                                        next_nodes[key] = [vals]
+                                elif key in later_nodes:
+                                    later_nodes[key].append(vals)
```

Comment:
Appears to be untested.

Reply:
Now covered after switching to `defaultdict(list)`